src/codegen/solana_accounts.rs

```diff
@@ -101,44 +87,110 @@ struct RecurseData<'a> {
     contract_no: usize,
     /// The quantity of accounts we have added the the hashmap 'accounts'
     accounts_added: usize,
+    /// The number of the AST function we are currently traversing
+    ast_no: usize,
     ns: &'a Namespace,
+    /// The vector of functions from the contract
+    functions: Vec<Function>,
 }
 
 impl RecurseData<'_> {
-    /// Add an account to the hashmap
-    fn add_account(&mut self, account: SolanaAccount) {
-        let inserted = match self.accounts.entry(self.cfg_func_no) {
-            Entry::Occupied(mut val) => val.get_mut().insert(account),
-            Entry::Vacant(val) => {
-                val.insert(IndexSet::from([account]));
-                true
-            }
-        };
-
-        if inserted {
+    /// Add an account to the function's indexmap
+    fn add_account(&mut self, account_name: String, account: SolanaAccount) {
+        if self.functions[self.ast_no]
+            .solana_accounts
+            .insert(account_name, account)
+            .is_none()
+        {
             self.accounts_added += 1;
         }
     }
+
+    /// Add the system account to the function's indexmap
+    fn add_system_account(&mut self) {
+        self.add_account(
+            SYSTEM_ACCOUNT.to_string(),
+            SolanaAccount {
+                is_writer: false,
+                is_signer: false,
+            },
+        );
+    }
 }
 
 /// Collect the accounts this contract needs
-pub(super) fn collect_accounts_from_contract(
-    contract_no: usize,
-    ns: &Namespace,
-) -> HashMap<usize, IndexSet<SolanaAccount>> {
-    let mut visiting_queue: IndexSet<(usize, usize)> = IndexSet::from_iter(
-        ns.contracts[contract_no]
-            .functions
-            .iter()
-            .map(|ast_no| (contract_no, ns.contracts[contract_no].all_functions[ast_no])),
-    );
+pub(super) fn collect_accounts_from_contract(contract_no: usize, ns: &mut Namespace) {
+    let mut functions: Vec<Function> = Vec::new();
+    std::mem::swap(&mut functions, &mut ns.functions);
```

Comment:
This is not great, but having said that I don't know how to write this any better. @xermicus have you got any ideas?

Reply:
I need to modify `ns.functions`, but I also need a reference to `ns` inside `RecurseData` and I already borrow `ns` to traverse the CFG, so rust complains about borrowing `ns` again as mutable. Swapping functions twice was the solution I found.

Reply:
A I see, maybe collect + assign the result would solve it?

Reply:
> A I see, maybe collect + assign the result would solve it?

A collect would involve copies, wouldn't it?

Reply:
I doubt this will have a noticeable performance impact. I favor easier code over premature optimizations here.

Reply:
I agree with @xermicus, don't optimize for execution speed over code clarity.

You're assuming the copy is of `struct Function`, but I was thinking the collect and assign could be the solana accounts, not the entire `Function`.

Either way, I don't think there are perfect solutions here and I am not sure how useful it is to spend more time on this.